shared-module/audiofilters/Distortion.c

```diff
@@ -0,0 +1,374 @@
+#include "shared-bindings/audiofilters/Distortion.h"
+#include "shared-module/audiofilters/Distortion.h"
+
+/**
+ * Based on Godot's AudioEffectDistortion
+ * - https://docs.godotengine.org/en/stable/classes/class_audioeffectdistortion.html
+ * - https://github.com/godotengine/godot/blob/master/servers/audio/effects/audio_effect_distortion.cpp
+ */
+
+void common_hal_audiofilters_distortion_construct(audiofilters_distortion_obj_t *self,
+    mp_obj_t drive, mp_obj_t pre_gain, mp_obj_t post_gain,
+    audiofilters_distortion_mode mode, mp_obj_t mix,
+    uint32_t buffer_size, uint8_t bits_per_sample, bool samples_signed,
+    uint8_t channel_count, uint32_t sample_rate) {
+
+    // Basic settings every effect and audio sample has
+    // These are the effects values, not the source sample(s)
+    self->bits_per_sample = bits_per_sample; // Most common is 16, but 8 is also supported in many places
+    self->samples_signed = samples_signed; // Are the samples we provide signed (common is true)
+    self->channel_count = channel_count; // Channels can be 1 for mono or 2 for stereo
+    self->sample_rate = sample_rate; // Sample rate for the effect, this generally needs to match all audio objects
+
+    // To smooth things out as CircuitPython is doing other tasks most audio objects have a buffer
+    // A double buffer is set up here so the audio output can use DMA on buffer 1 while we
+    // write to and create buffer 2.
+    // This buffer is what is passed to the audio component that plays the effect.
+    // Samples are set sequentially. For stereo audio they are passed L/R/L/R/...
+    self->buffer_len = buffer_size; // in bytes
+
+    self->buffer[0] = m_malloc(self->buffer_len);
+    if (self->buffer[0] == NULL) {
+        common_hal_audiofilters_distortion_deinit(self);
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[0], 0, self->buffer_len);
+
+    self->buffer[1] = m_malloc(self->buffer_len);
+    if (self->buffer[1] == NULL) {
+        common_hal_audiofilters_distortion_deinit(self);
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[1], 0, self->buffer_len);
+
+    self->last_buf_idx = 1; // Which buffer to use first, toggle between 0 and 1
+
+    // Initialize other values most effects will need.
+    self->sample = NULL; // The current playing sample
+    self->sample_remaining_buffer = NULL; // Pointer to the start of the sample buffer we have not played
+    self->sample_buffer_length = 0; // How many samples do we have left to play (these may be 16 bit!)
+    self->loop = false; // When the sample is done do we loop to the start again or stop (e.g. in a wav file)
+    self->more_data = false; // Is there still more data to read from the sample or did we finish
+
+    // The below section sets up the effect's starting values.
+
+    // If we did not receive a BlockInput we need to create a default float value
+    if (drive == MP_OBJ_NULL) {
+        drive = mp_obj_new_float(0.0);
```

Comment:
Handle default float values that happen to be integers by using MP_ROM_INT, so this check-and-assign can be removed.

Reply:
Yes, I'll be following your recent examples on this as well as removing the following unnecessary kwarg setters:
```
audiofilters_distortion_obj_set_drive
audiofilters_distortion_obj_set_pre_gain
audiofilters_distortion_obj_set_post_gain
audiofilters_distortion_obj_set_mode
audiofilters_distortion_obj_set_mix
```